### PR TITLE
Dev/add memo to token creation

### PR DESCRIPTION
**Pull request by drbh**
- Added a new flag to add a memo on `create-token`
- Add a transaction instruction to write memo via `Memo Program v2`

The `Vec<u8>` address was derived from the `devnet` existing program address (there probably a better way to import the program id based on the network (not sure how to retrieve those values)

```python3
>>> import base58
>>> bytelist = base58.b58decode("MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr")
>>> barr = [b for b in bytelist]
>>> barr
# [5, 74, 83, 90, 153, 41, 33, 6, 77, 36, 232, 113, 96, 218, 56, 124,
# 124, 53, 181, 221, 188, 146, 187, 129, 228, 31, 168, 64, 65, 5, 68, 141]
```

### CODE REVIEW

**Comment by mvines** [comment]
Cool, you can just import the spl_memo crate and use the id from there.  It's the same id across all clusters.

Here's a similar example from when memo support was added to the main cli: https://github.com/solana-labs/solana/pull/16291/files

File: token/cli/Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ solana-remote-wallet = "=1.6.7"
 solana-sdk = "=1.6.7"
 spl-token = { version = "3.1", path="../program", features = [ "no-entrypoint" ] }
 spl-associated-token-account = { version = "1.0", path="../../associated-token-account/program", features = [ "no-entrypoint" ] }
+spl-memo = { version = "=3.0.1", features = ["no-entrypoint"] }
```
Comment (mvines) [inline token/cli/Cargo.toml:25]: please just path out to the spl-memo that's already in the tree, like we do for spl-token above

File: token/cli/src/main.rs [hunk 2]
```diff
@@ -254,6 +256,25 @@ fn command_create_token(
             decimals,
         )?,
     ];
+
+    // check for memo text; if it exsits convert it to
```
Comment (mvines) [inline token/cli/src/main.rs:260]:
```suggestion
    // check for memo text; if it exists convert it to
```

File: token/cli/src/main.rs [hunk 3]
```diff
@@ -2051,6 +2078,11 @@ fn main() {
     let _ = match (sub_command, sub_matches) {
         ("create-token", Some(arg_matches)) => {
             let decimals = value_t_or_exit!(arg_matches, "decimals", u8);
+            let memo = if arg_matches.is_present("memo") {
+                Some(value_t_or_exit!(arg_matches, "memo", String))
+            } else {
+                None
+            };
```
Comment (mvines) [inline token/cli/src/main.rs:2085]:
```suggestion
            let memo = value_t!(arg_matches, "memo", String).ok();
```

**Comment by mvines** [comment]
Thanks!  I made a couple comments.  Please ping me again after clippy passes and I'll be happy to review the PR again

File: token/cli/src/main.rs [hunk 4]
```diff
@@ -254,6 +255,21 @@ fn command_create_token(
             decimals,
         )?,
     ];
+
+    // check for memo text; if it exists convert it to
+    // bytes and add memo tx instructions
+    if let Some(text) = memo {
+        let memo_instruction = Instruction {
```
Comment (CriesofCarrots) [inline token/cli/src/main.rs:262]: Can we use `spl_memo::build_memo()` to save some code here? https://github.com/solana-labs/solana-program-library/blob/731bb6f57a676c2ad8c7a8ed2ea094f32d83cc5f/memo/program/src/lib.rs#L30

**Comment by drbh** [comment]
@mvines updates made and clippy passed, please let me know if I should make any more changes. Thanks

File: token/cli/src/main.rs [hunk 5]
```diff
@@ -254,6 +255,14 @@ fn command_create_token(
             decimals,
         )?,
     ];
+
+    // check for memo text; if it exists convert it to
+    // bytes and add memo tx instructions
+    if let Some(text) = memo {
+        let memo_instruction = spl_memo::build_memo(text.as_bytes(), &[&config.owner]);
+        instructions.push(memo_instruction);
```
Comment (mvines) [inline token/cli/src/main.rs:263]: how about a one-liner here:
```suggestion
        instructions.push(spl_memo::build_memo(text.as_bytes(), &[&config.owner]));
```

**Review by mvines — COMMENTED** [review]
Looks good!  Just one nit, and then github is complaining that it can't rebase the commits in this PR.  Can you please squash them all down to 1 commit and then rebase that 1 commit on the HEAD of the master branch.  That should make github happy

**Comment by drbh** [comment]
closing in favor of newer PR https://github.com/solana-labs/solana-program-library/pull/1709

**Pull request closed without merging**